src/plugins/WebOfTrust/util/DelayedBackgroundJob.java

```diff
@@ -0,0 +1,270 @@
+package plugins.WebOfTrust.util;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import freenet.node.FastRunnable;
+import freenet.node.PrioRunnable;
+import freenet.support.Executor;
+import freenet.support.Ticker;
+import freenet.support.io.NativeThread;
+
+/**
+ * A {@link BackgroundJob} that is executed after triggers have been aggregated for a given delay.
+ *
+ * @author bertm
+ * @see BackgroundJob
+ * @see DelayedBackgroundJobFactory
+ */
+public class DelayedBackgroundJob implements BackgroundJob {
+    static enum JobState {
+        /** Waiting for a trigger, no running job thread or scheduled job. */
+        IDLE,
+        /** Waiting for the scheduled job to be executed, no running job thread. */
+        WAITING,
+        /** Running the job in the job thread. */
+        RUNNING,
+        /** Waiting for the running job thread to finish before we terminate. */
+        TERMINATING,
+        /** Terminated, no running job thread. */
+        TERMINATED
+    }
+
+    /** Special time value to indicate the absence of an oldest trigger. */
+    private final static long NO_TRIGGER = 0;
+
+    /** The executor of this background job. */
+    private final Executor executor;
+    /** The ticker used to schedule this job. */
+    private final Ticker ticker;
+    /** Job wrapper for status tracking. */
+    private final DelayedBackgroundRunnable realJob;
+    /** Human-readable name of this job. */
+    private final String name;
+    /** Aggregation delay in milliseconds. */
+    private final long delay;
+
+    /** The time at which the first trigger occurred after the latest job execution start. */
+    private final AtomicLong oldestTrigger = new AtomicLong(NO_TRIGGER);
+    /** Running state of this job. */
+    private JobState state = JobState.IDLE;
+    /** Job execution thread, only if the job is running. */
+    private Thread thread = null;
+
+    /**
+     * Constructs a delayed background job.
+     * The {@link Executor} and {@link Ticker} given <b>must</b> have an asynchronous implementation
+     * of respectively {@link Executor#execute(Runnable, String) execute} and
+     * {@link Ticker#queueTimedJob(Runnable, String, long, boolean, boolean) queueTimedJob}.
+     * @param job the job to run in the background
+     * @param name a human-readable name for the job
+     * @param delay the background job aggregation delay in milliseconds
+     * @param executor an asynchronous executor
+     * @param ticker an asynchronous ticker
+     */
+    DelayedBackgroundJob(Runnable job, String name, long delay, Executor executor, Ticker ticker) {
```

Comment:
> I'd rather warn the implementer in the javadoc instead of forcing him to implement another interface on his runnables. 

There still would only be *one* run function because adding another exception to the throws declaration does not make it a different function. To the reader of such classes, it would have the benefit of being obvious that the Runnable can be aborted from just reading the function signature. Otherwise, one has to dig into the code.

> Regarding the constructor: it's perfectly valid to use the constructor directly, you just miss out on the (relatively minor) benefits of the factory. There's no reason to make the constructor package-private and recommend the factory; the factory performs no additional validation. I will add the annotation to point to the existence of the factory, however.

Not needing to use the factory is actually a nice thing. I had already feared that they non-necessarily bloat the WOT code :) I would suggest you explicitly state at the JavaDoc of the constructors and of the factories that they don't have to be used.